### PR TITLE
Faster dot product

### PR DESCRIPTION
**Pull request by mundya**
Uses the ARM `smlal` instruction to improve performance of neuron update step.
- [x] Fit model from IJCNN paper
  - [x] Update partitioning to account for this change in cost

These graphs are comparable to those in the IJCNN paper: the left hand bars show current master, the right hand bars show for this PR.

profile_16_dimensions
profile_100_neurons

### CODE REVIEW

**Comment by mundya** [comment]
After 9725565: (bars in order: master, after a06676e, after 9725565)
profile_16_dimensions
profile_100_neurons

**Comment by neworderofjamie** [comment]
Hah, take that less than operator :-/

**Comment by mundya** [comment]
At this point we're here vs master (left hand bar is master, right hand is now).

profile_16_dimensions
profile_100_neurons

I slightly changed to model I was profiling against to give a better indication of reasonable firing rates.  That said there's still some space for someone to look into the distribution of neuron update costs.

update_costs

**Comment by mundya** [comment]
For reference, profiling model:

``` python
with nengo.Network() as model:
    # Create the input to the network.
    ins = [nengo.Node(WhiteNoise(), label="inp") for _ in range(n_dims)]

    # Create the ensemble which we'll be profiling
    a = nengo.Ensemble(n_neurons, n_dims, label="A")

    # Create an ensemble to act as the end of the communication channel
    b = nengo.Ensemble(n_neurons, n_dims, label="B")

    # Connect the input to the first ensemble, and the first ensemble to
    # the second.
    for i, inp in enumerate(ins):
        nengo.Connection(inp, a[i], transform=1.0/np.sqrt(n_dims))
    nengo.Connection(a, b)

# Specify that we want to generate the input on chip and that we want to
# profile one of the ensembles.
nengo_spinnaker.add_spinnaker_params(model.config)
for node in model.all_nodes:
    model.config[node].function_of_time = True
model.config[a].profile = True
```

**Comment by mundya** [comment]
OK, profiling!

| Cost | IJCNN Paper | Master | This PR |
| --- | --- | --- | --- |
| Input filtering | 43D + 245 | 42D + 181 | 40D + 196 |
| Neuron update | 13ND + 69N + 188 | 15ND + 88N | 9ND + 69N |
| Transmit output | 702D + 100 | 123D + 46 | 125D + 45 |

Where N is the number of neurons and D is the number of dimensions.
- _Input filtering_ Has changed little, as expected.
- _Neuron update_
  - I've removed the constant term as it wasn't significant
  - Between Master and This PR the neuron update cost decreases significantly
  - Between IJCNN and this PR the neuron model was improved and voltage recording added
- _Transmit output_ between IJCNN and Master I removed a 1us delay between transmitting each packet, hence the huge reduction in compute time.

**Comment by mundya** [comment]
As a side note: that's around 2500 neurons per core in 1D or (the sadly unreachable due to memory boundedness) 925 neurons per core in 16D.

**Comment by tcstewar** [comment]
> 925 neurons per core in 16D.

Awesome.  :) :)

File: spinnaker_components/common/fixed_point.h [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+#include "arm_acle_gcc_selected.h"
+#include "nengo_typedefs.h"
+
+#ifndef __FIXED_POINT_H__
+#define __FIXED_POINT_H__
+
+/*****************************************************************************/
+// Convert from an INT64 to appropriate INT32 fixed point representation
+
+static inline int32_t scale_64_to_32(int64_t value)
```
Comment (neworderofjamie) [inline spinnaker_components/common/fixed_point.h]: This is a bit ambiguous  - convert_s32_30_s16_15 is also gross, but maybe more descriptive..

File: spinnaker_components/common/fixed_point.h [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    //   {
+    //     result32 = -INT32_MAX;
+    //   }
+
+    // The sign bit should be set as this value is negative.
+    result32 |= (1 << 31);
+  }
+
+  // Return the fixed point value
+  return result32;
```
Comment (neworderofjamie) [inline spinnaker_components/common/fixed_point.h]: What happens if you just shift down the 64-bit value and cast?

File: spinnaker_components/ensemble/ensemble_update.c [hunk 3]
```diff
@@ -77,23 +78,45 @@ void ensemble_update(uint ticks, uint arg1) {
     // Save state
     set_neuron_voltage(n, v_voltage);
 
-    // If this neuron has fired then process
-    if( v_voltage > 1.0k ) {
-      // Set the voltage to be the overshoot, set the refractory time
-      set_neuron_refractory(n);
-      set_neuron_voltage(n, v_voltage - 1.0k);
+    if (bitsk(v_voltage) <= bitsk(1.0k))
```
Comment (neworderofjamie) [inline spinnaker_components/ensemble/ensemble_update.c:84]: Add comment as to the horror

**Comment by mundya** [comment]
All done.

**Comment by mundya** [comment]
Merging after checking with @neworderofjamie.

**Pull request merged**